beaker_kernel/lib/subkernels/python.py

```diff
@@ -52,3 +55,51 @@ def parse_subkernel_return(cls, execution_result) -> Any:
         if return_str:
             python_obj = ast.literal_eval(return_str)
             return python_obj
+
+    async def generate_checkpoint_from_state(self) -> Checkpoint:
+        save_state_code = """
+import inspect as _inspect
+import json as _json
+import dill as _dill
+class _SubkernelStateEncoder(_json.JSONEncoder):
+    def default(self, o):
+        # if callable(o):
+            # return f"Function named"
+            # return super().default(o)
+        try:
+            return super().default(o)
+        except:
+            return str(o)
+
+_result = {}
+for _name, _value in dict(locals()).items():
+    if _name.startswith('_') or _name in ('In', 'Out', 'get_ipython', 'exit', 'quit', 'open'):
+        continue
+    _path = f"%s/{_name}.pkl"
+    with open(_path, "wb") as _f:
+       _dill.dump(_value, _f)
+       _result[_name] = _path
+
+_result = _json.loads(_json.dumps(_result, cls=_SubkernelStateEncoder))
+del _inspect, _json, _dill
+_result
+""" % self.storage_prefix
+        response = await self.evaluate(save_state_code)
+        return response["return"]
+        
+
+    async def load_checkpoint(self, checkpoint: Checkpoint):
+        vars = await self.evaluate("""
+import dill as _dill
+_exclusion_critieria = lambda name: name.startswith('_') or name in ('In', 'Out', 'get_ipython', 'exit', 'quit', 'open')
+[ name for name, value in dict(locals()).items() if not _exclusion_critieria(name) ]
+""")
+        await self.context.evaluate(f"del {', '.join(vars['return'])}")
+        for varname, filename in checkpoint.items():
+            load_state_code = f"""
+with open("{filename}", "rb") as _file:
+    {varname} = _dill.load(_file)
+""" 
+            await self.evaluate(load_state_code)
```

Comment:
Probably would be more efficient if we concatenated the commands here and then executed (not evaluate) once.

Evaluate is for when you want to retrieve the response so that you can work with it in kernelspace and execute is when you just need to run something and don't need to use the output.